extensions-core/kafka-indexing-service/src/main/java/org/apache/druid/data/input/kafkainput/KafkaInputReader.java

```diff
@@ -0,0 +1,211 @@
+
+  private final InputRowSchema inputRowSchema;
+  private final KafkaRecordEntity record;
+  private final KafkaHeaderReader headerParser;
+  private final InputEntityReader keyParser;
+  private final InputEntityReader valueParser;
+  private final String keyLabelPrefix;
+  private final String recordTimestampLabelPrefix;
+
+  public KafkaInputReader(
+      InputRowSchema inputRowSchema,
+      KafkaRecordEntity record,
+      KafkaHeaderReader headerParser,
+      InputEntityReader keyParser,
+      InputEntityReader valueParser,
+      String keyLabelPrefix,
+      String recordTimestampLabelPrefix
+  )
+  {
+    this.inputRowSchema = inputRowSchema;
+    this.record = record;
+    this.headerParser = headerParser; //Header parser can be null by config
+    this.keyParser = keyParser; //Key parser can be null by config and data
+    this.valueParser = valueParser; //value parser can be null by data (tombstone records)
+    this.keyLabelPrefix = keyLabelPrefix;
+    this.recordTimestampLabelPrefix = recordTimestampLabelPrefix;
+  }
+
+  @Override
+  public CloseableIterator<InputRow> read() throws IOException
+  {
+    // Add kafka record timestamp to the mergelist
+    List<Pair<String, Object>> mergeList = new ArrayList<>();
+    if (headerParser != null) {
+      List<Pair<String, Object>> headerList = headerParser.read();
+      mergeList.addAll(headerList);
+    }
+
+    Pair ts = new Pair(recordTimestampLabelPrefix + DEFAULT_TIMESTAMP_STRING,
+                      record.getRecord().timestamp());
+    if (!mergeList.contains(ts)) {
+      mergeList.add(ts);
+    }
+
+    if (keyParser != null) {
+      try (CloseableIterator<InputRow> keyIterator = keyParser.read()) {
+        // Key currently only takes the first row and ignores the rest.
+        if (keyIterator.hasNext()) {
+          // Return type for the key parser should be of type MapBasedInputRow
+          // Parsers returning other types are not compatible currently.
+          MapBasedInputRow keyRow = (MapBasedInputRow) keyIterator.next();
+          Pair key = new Pair(keyLabelPrefix + DEFAULT_KEY_STRING,
+                            keyRow.getEvent().entrySet().stream().findFirst().get().getValue());
+          if (!mergeList.contains(key)) {
+            mergeList.add(key);
+          }
+        }
+      }
+      catch (IOException e) {
+        throw e;
+      }
+      catch (ClassCastException e) {
+        log.error(e, "Encountered ClassCastException exception, please use parsers that return MapBasedInputRow based rows.");
+      }
+    }
+
+    if (valueParser != null) {
+      CloseableIterator<InputRow> iterator = valueParser.read();
+
+      return new CloseableIterator<InputRow>()
+      {
+        @Override
+        public boolean hasNext()
+        {
+          return iterator.hasNext();
+        }
+
+        @Override
+        public InputRow next()
+        {
+          InputRow row = null;
```

Comment:
this can be removed if you return directly from your try block on successful parsing.